common/include/verilog_int.h

```diff
@@ -450,38 +466,82 @@ struct vint {
 	//////////////////////
 	// shift assignment
 	//////////////////////
-	/*
-	vint& operator>>=(const unsigned rhs) {
+	void signext_after_shiftr(const unsigned rhs) {
+		const unsigned num_remaining = num_bit - rhs;
+		const unsigned signext_from_word = detail::num_bit2num_word(num_remaining);
+		const unsigned used_bit_after_shift_minus_one = (num_remaining-1u) % bw_word;
+		if constexpr (num_word > 1) {
+			assert(signext_from_word > 0); // since rhs != 0, this shall not fail
+			::std::fill(v+signext_from_word, v+num_word, stype(-1));
+		}
+		v[signext_from_word-1] |= stype(-2) << used_bit_after_shift_minus_one;
+		ClearUnusedBits();
+	}
+
+	void clear_after_shiftr(const unsigned rhs) {
+		if constexpr (num_word > 1) {
+			const unsigned num_remaining = num_bit - rhs;
+			const unsigned signext_from_word = detail::num_bit2num_word(num_remaining);
+			const unsigned unused_bit_after_shift = (-num_remaining) % bw_word;
+			assert(signext_from_word > 0); // since rhs != 0, this shall not fail
+			::std::fill(v+signext_from_word, v+num_word, stype(0));
+			v[signext_from_word-1] &= stype(-1) >> unused_bit_after_shift;
+		}
+	}
+
+	[[gnu::noinline]]
+	void raw_shiftopr(const unsigned rhs, bool is_neg) {
 		// rhs >= num_bit is UB in C, but not sure in Verilog
 		if (rhs == 0 or rhs >= num_bit) {
 			// do nothing
-			return *this;
+			return;
 		}
-
 		if constexpr (num_word == 1) {
 			v[0] >>= rhs;
 		} else {
-			const dtype sign_ext = v[num_word-1] >> 63;
 			const unsigned word_shift = rhs / 64;
 			const unsigned bit_shift = rhs % 64;
 			if (bit_shift == 0) {
 				for (unsigned i = word_shift; i < num_word; ++i) {
 					v[i-word_shift] = v[i];
 				}
 			} else {
-				const unsigned interleave_bit = bit_shift;
 				for (unsigned i = word_shift; i < num_word-1; ++i) {
-					v[i-word_shift] = detail::interleave64(v[i], v[i+1], interleave_bit);
+					v[i-word_shift] = detail::shiftright128(v[i+1], v[i], bit_shift);
 				}
-				v[num_word-word_shift-1] = detail::interleave64(v[num_word-1], sign_ext, interleave_bit);
+				v[num_word-word_shift-1] = v[num_word-1] >> bit_shift;
 			}
-			::std::fill_n(::std::end(v) - word_shift, word_shift, sign_ext);
 		}
-		// Not necessary for right shift
-		// ClearUnusedBits();
+		if (is_neg) {
+			signext_after_shiftr(rhs);
+		} else {
+			clear_after_shiftr(rhs);
+		}
+	}
+
+	// >>=, unsigned
+	vint& ushiftopr(const unsigned rhs) {
+		raw_shiftopr(rhs, false);
+		return *this;
+	}
+
+	// >>=, signed
+	vint& sshiftopr(const unsigned rhs) {
+		const bool is_neg = Bit(num_bit-1u);
+		raw_shiftopr(rhs, is_neg);
+		return *this;
+	}
+
+	vint& operator>>=(const unsigned rhs) {
+		if constexpr (is_signed) {
```

Comment:
How about just a one `shiftopr`, with argument is_signed.
Then `is_neg = is_signed && Bit(num_bit-1)`

Reply:
I think this is good. Let me unify the shiftopr/assign/uvalue